File: spring-data/src/main/java/dev/cerbos/queryplan/springdata/TernaryTranslator.java
/*
 * Copyright 2021-2026 Zenauth Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

package dev.cerbos.queryplan.springdata;

import dev.cerbos.api.v1.engine.Engine.PlanResourcesFilter;
import dev.cerbos.api.v1.engine.Engine.PlanResourcesFilter.Expression.Operand;

import jakarta.persistence.criteria.CriteriaBuilder;
import jakarta.persistence.criteria.Predicate;

import java.util.List;
import java.util.function.Function;

/**
 * Translates the CEL ternary {@code if(c, a, b)} as a whole condition
 * ({@link #handleBareTernary}) and as a comparison operand ({@link #tryTernaryComparison}).
 *
 * <p>Each branch is substituted back into the surrounding shape and walked again, so it
 * translates exactly like the same condition written directly.
 */
final class TernaryTranslator {

    private final CriteriaBuilder cb;
    private final TriPredicate tri;
    private final PlanWalker walker;

    TernaryTranslator(CriteriaBuilder cb, TriPredicate tri, PlanWalker walker) {
        this.cb = cb;
        this.tri = tri;
        this.walker = walker;
    }

    /**
     * Rewrites {@code cmp(if(c, a, b), other)} as {@code cmp(a, other)} and
     * {@code cmp(b, other)}. Runs on the raw operands, before any mirroring. Nested ternaries
     * and a ternary on the other side are handled by the recursion.
     *
     * @return the rewritten predicate, or {@code null} if this comparison has no ternary
     */
    Predicate tryTernaryComparison(String op, List<Operand> operands, Scope scope) {
        if (!ComparisonTranslator.COMPARISON_OPS.contains(op) || operands.size() != 2) {
            return null;
        }
        int idx;
        if (isIfExpression(operands.get(0))) {
            idx = 0;
        } else if (isIfExpression(operands.get(1))) {
            idx = 1;
        } else {
            return null;
        }
        List<Operand> ifOps = operands.get(idx).getExpression().getOperandsList();
        return translateTernary(ifOps,
                branch -> walker.traverseExpression(substituteOperand(op, operands, idx, branch), scope),
                scope);
    }

    private static boolean isIfExpression(Operand o) {
        return o.getNodeCase() == Operand.NodeCase.EXPRESSION
                && "if".equals(o.getExpression().getOperator());
    }

    /**
     * Rewrites {@code if(c, a, b)} with {@link TriPredicate#ternary}, which keeps the result
     * UNKNOWN when {@code c} is (CEL denies a null condition). A predicate rewrite rather than
     * {@code CASE WHEN} sends each branch through the normal leaf paths.
     *
     * <p>A constant condition translates only the live branch, so an untranslatable dead branch
     * cannot fail the plan.
     */
    private Predicate translateTernary(List<Operand> ifOps,
                                       Function<Operand, Predicate> branchTranslator,
                                       Scope scope) {
        if (ifOps.size() != 3) {
            throw Refusals.malformed(
                    "if (ternary) requires exactly 3 operands (condition, then, else), got "
                            + ifOps.size());
        }
        Operand condition = ifOps.get(0);
        Operand thenBranch = ifOps.get(1);
        Operand elseBranch = ifOps.get(2);

        if (condition.getNodeCase() == Operand.NodeCase.VALUE) {
            Boolean known = constantBooleanOrNull(condition);
            if (known == null) {
                // A non-boolean literal condition is a CEL type error.
                throw Refusals.malformed(
                        "if (ternary) condition must be a boolean expression");
            }
            return branchTranslator.apply(known ? thenBranch : elseBranch);
        }

        return tri.ternary(
                () -> walker.traverse(condition, scope),
                () -> branchTranslator.apply(thenBranch),
                () -> branchTranslator.apply(elseBranch));
    }

    /** {@code if(c, a, b)} used directly as a condition, so both branches are boolean. */
    Predicate handleBareTernary(List<Operand> operands, Scope scope) {
        return translateTernary(operands, branch -> booleanBranchPredicate(branch, scope), scope);
    }

    // A boolean constant branch becomes always-true or always-false.
    private Predicate booleanBranchPredicate(Operand branch, Scope scope) {
        if (branch.getNodeCase() == Operand.NodeCase.VALUE) {
            Boolean constant = constantBooleanOrNull(branch);
            if (constant == null) {
                throw Refusals.malformed(
                        "if (ternary) branch in boolean position must be a boolean");
            }
            return constant ? cb.conjunction() : cb.disjunction();
        }
        return walker.traverse(branch, scope);
    }

    private static PlanResourcesFilter.Expression substituteOperand(
            String op, List<Operand> operands, int idx, Operand replacement) {
        PlanResourcesFilter.Expression.Builder b =
                PlanResourcesFilter.Expression.newBuilder().setOperator(op);
        for (int i = 0; i < operands.size(); i++) {
            b.addOperands(i == idx ? replacement : operands.get(i));
        }
        return b.build();
    }

    private static Boolean constantBooleanOrNull(Operand o) {
        return PlanValues.protoValueToJava(o.getValue()) instanceof Boolean b ? b : null;
    }
}

File: exposed/src/main/kotlin/dev/cerbos/queryplan/exposed/TernaryTranslator.kt
package dev.cerbos.queryplan.exposed

import dev.cerbos.api.v1.engine.Engine.PlanResourcesFilter
import dev.cerbos.api.v1.engine.Engine.PlanResourcesFilter.Expression.Operand
import org.jetbrains.exposed.v1.core.Op

/**
 * The CEL ternary, `if(condition, whenTrue, whenFalse)`, in both positions. Both are REWRITES, not
 * lowerings to `CASE`: the branches are substituted back into the surrounding shape and walked
 * again, so a branch translates exactly as the same condition written directly, and the three
 * predicates are combined with [TriLogic.ternary].
 *
 * Rewriting rather than emitting `CASE WHEN` is what keeps every typed leaf path — field-first
 * normalisation, the `size()` handling, the `add` fold, double-space comparison — reachable from
 * inside a ternary. Nesting, and a ternary on the other side of the comparison, then come for
 * free through the recursion.
 */
internal class TernaryTranslator(private val translation: Translation) {

    /** `if(...)` as the whole condition: both branches are themselves boolean. */
    fun translateBare(operands: List<Operand>, scope: Scope): Op<Boolean> =
        translateTernary(operands, scope) { branch -> translation.walker.traverse(branch, scope) }

    /**
     * A comparison with a ternary operand, or `null` when neither operand is one. Called FIRST by
     * [ComparisonTranslator.translate], on raw operands, because it must see source order: each
     * branch is substituted back into the comparison AS WRITTEN and walked again, and mirroring
     * first would substitute into a comparison the policy never had.
     */
    fun tryTernaryComparison(operator: String, operands: List<Operand>, scope: Scope): Op<Boolean>? {
        if (operator !in ComparisonTranslator.COMPARISON_OPERATORS || operands.size != 2) return null
        val index = when {
            isTernary(operands[0]) -> 0
            isTernary(operands[1]) -> 1
            else -> return null
        }
        return translateTernary(operands[index].expression.operandsList, scope) { branch ->
            translation.walker.traverseExpression(substitute(operator, operands, index, branch), scope)
        }
    }

    /**
     * The rewrite itself. A CONSTANT boolean condition folds to a single branch, and only that
     * branch is translated — so an untranslatable dead branch cannot fail the whole plan.
     *
     * Null semantics and the third arm belong to [TriLogic.ternary]: a NULL condition column is a
     * CEL evaluation error, which denies, so the predicate has to be UNKNOWN there rather than
     * FALSE — two false branches would otherwise collapse it to FALSE and an enclosing negation
     * would readmit the row.
     */
    private fun translateTernary(
        operands: List<Operand>,
        scope: Scope,
        branch: (Operand) -> Op<Boolean>,
    ): Op<Boolean> {
        if (operands.size != 3) {
            throw Refusals.malformed(
                "if (ternary) requires exactly 3 operands (condition, then, else), got ${operands.size}",
            )
        }
        val condition = operands[0]
        if (condition.nodeCase == Operand.NodeCase.VALUE) {
            val known = PlanValues.toKotlin(condition.value) as? Boolean
                // A non-boolean literal condition is a CEL type error the planner never emits.
                ?: throw Refusals.malformed("if (ternary) condition must be a boolean expression")
            return branch(operands[if (known) 1 else 2])
        }
        return TriLogic.ternary(
            translation.walker.traverse(condition, scope),
            branch(operands[1]),
            branch(operands[2]),
        )
    }

    private fun isTernary(operand: Operand): Boolean =
        operand.nodeCase == Operand.NodeCase.EXPRESSION && operand.expression.operator == "if"

    /** Rebuilds `operator(operands...)` with the operand at [index] replaced. */
    private fun substitute(
        operator: String,
        operands: List<Operand>,
        index: Int,
        replacement: Operand,
    ): PlanResourcesFilter.Expression {
        val builder = PlanResourcesFilter.Expression.newBuilder().setOperator(operator)
        operands.forEachIndexed { position, operand ->
            builder.addOperands(if (position == index) replacement else operand)
        }
        return builder.build()
    }
}
